src/tree/utils.ts

```diff
@@ -1,9 +1,23 @@
-import { isReferenceNode, isRegularNode, SchemaNode, SchemaNodeKind } from '@stoplight/json-schema-tree';
+import {
+  isReferenceNode,
+  isRegularNode,
+  MirroredSchemaNode,
+  ReferenceNode,
+  RegularNode,
+  SchemaNode,
+  SchemaNodeKind,
+} from '@stoplight/json-schema-tree';
 
 import { isNonNullable } from '../guards/isNonNullable';
 import { ComplexArrayNode, FlattenableNode, PrimitiveArrayNode } from './types';
 
-export const isParentNode = (node: SchemaNode) => isRegularNode(node) && !!node.children && node.children.length > 0;
+export type ChildNode = RegularNode | ReferenceNode | MirroredSchemaNode;
+
+export const isNonEmptyParentNode = (
+  node: SchemaNode,
+): node is RegularNode & {
+  children: ChildNode[] & { 0: ChildNode };
```

Comment:
lol. This blew my mind. :joy:

Reply:
Yeah, unfortunately TS does not have any other way to accomplish this so we end up using that in a few projects :laughing:

Reply:
> *praying no one mentions the word Monad 🙏*